lib/prop/leaky_bucket_strategy.rb

```diff
@@ -5,40 +5,73 @@
 module Prop
   class LeakyBucketStrategy
     class << self
+      def _throttle_leaky_bucket(handle, key, cache_key, options)
+        (over_limit, bucket) = options.key?(:decrement) ?
+          decrement(cache_key, options.fetch(:decrement), options) :
+          increment(cache_key, options.fetch(:increment, 1), options)
+
+        if over_limit
+          [true, bucket]
+        else
+          [false, bucket]
+        end
+      end
+
       def counter(cache_key, options)
-        bucket = Prop::Limiter.cache.read(cache_key) || zero_counter
-        now = Time.now.to_i
-        leak_rate = (now - bucket.fetch(:last_updated)) / options.fetch(:interval).to_f
-        leak_amount =  leak_rate * options.fetch(:threshold)
+        cache.read(cache_key) || zero_counter
+      end
+
+      def leak_amount(bucket, amount, options, now)
+        leak_rate = (now - bucket.fetch(:last_leak_time, 0)) / options.fetch(:interval).to_f
+        leak_amount = (leak_rate * options.fetch(:threshold).to_f)
+        leak_amount.to_i
+      end
 
-        bucket[:bucket] = [(bucket.fetch(:bucket) - leak_amount).to_i, 0].max
-        bucket[:last_updated] = now
-        bucket
+      def update_bucket(current_bucket_size, max_bucket_size, amount)
+        over_limit = (max_bucket_size-current_bucket_size) < amount
+        updated_bucket_size = over_limit ? current_bucket_size : current_bucket_size + amount
+        [over_limit, updated_bucket_size]
       end
 
       # WARNING: race condition
       # this increment is not atomic, so it might miss counts when used frequently
       def increment(cache_key, amount, options)
-        counter = counter(cache_key, options)
-        counter[:bucket] += amount
-        Prop::Limiter.cache.write(cache_key, counter)
-        counter
+        bucket = counter(cache_key, options)
+        now = Time.now.to_i
+        max_bucket_size = options.fetch(:burst_rate)
+        current_bucket_size = bucket.fetch(:bucket, 0)
+        leak_amount = leak_amount(bucket, amount, options, now)
+        if leak_amount > 0
+          # maybe TODO, update last_leak_time to reflect the exact time for the current leak amount
+          # the current strategy will always reflect a little less leakage, probably not an issue though
+          bucket[:last_leak_time] = now
+          current_bucket_size = [(current_bucket_size - leak_amount), 0].max
+        end
+
+        over_limit, updated_bucket_size = update_bucket(current_bucket_size, max_bucket_size, amount)
+        bucket[:bucket] = updated_bucket_size
+        bucket[:over_limit] = over_limit
+        cache_value = cache.write(cache_key, bucket)
```

Comment:
Do we need to save the value in a variable here?

Reply:
fixed